firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/ServerTimestampOperation.java

```diff
@@ -30,7 +30,10 @@ public static ServerTimestampOperation getInstance() {
   }
 
   @Override
-  public FieldValue applyToLocalView(@Nullable FieldValue previousValue, Timestamp localWriteTime) {
+  public FieldValue applyToLocalView(
+      @Nullable FieldValue currentValue,
+      @Nullable FieldValue previousValue,
+      Timestamp localWriteTime) {
     return new ServerTimestampValue(localWriteTime, previousValue);
```

Comment:
So this is the only implementation of applyToLocalView() that uses `previousValue` instead of `currentValue`, and I was surprised that it does...  It seems reasonable to expect that if you do:

```
db.batch()
  .update(ref, "foo", "old")
  .update(ref, "foo", FieldValue.serverTimestamp()
  .commit();
```

then the `ServerTimestampBehavior.PREVIOUS` value would be "old" (i.e. `currentValue`) rather than whatever the value was before the batch...

But I see if you change this to `currentValue` then we have a test that breaks...  But I think it breaks for a bad reason (a deeper bug in our code, possibly in the way `UserDataConverter` splits updates into a `PatchMutation` and a `TransformMutation`)

I would be a fan of:
1) Getting rid of `previousValue` and always using `currentValue`.  This will clean up the code and make it much easier to reason about.  FWIW I had a hard time groking `previousValue` and `currentValue` since they're both "previous" values.  More explicit names might be `preBatchValue` and `preMutationValue`...  This might technically be a breaking change for ServerTimestampBehavior.PREVIOUS in the case I demonstrated above, but I think it's probably subtle enough we don't care.
2) Tracking down / fixing whatever the bug that exposes. :-)  [sorry for the extra homework...]

Reply:
The problem the currentValue/previousValue "hack" addresses is actually different from what you suspect. A problematic call could be something simple as:

```
docRef.update(map("nested", map("time", FieldValue.serverTimestamp())
```

This turns into two mutations in a single mutation batch:

- A PatchMutation that replaces any existing value of "nested" with `map()`.
- A Transform mutation that sets `nested.time` to a ServerTimestamp.

By the time the Transform is applied, `maybeDoc` no longer contains the previous field value.

I added a comment to the test that explains this.

I do think we can improve our naming here, but I don't feel that re-working this entire setup is time well spent.